abci/app.go

```diff
@@ -0,0 +1,377 @@
+	fmt.Println("entered processProp")
+	bannedUsers := make(map[string]struct{}, 0)
+
+	finishedBanTxIdx := len(processproposal.Txs)
+	for i, tx := range processproposal.Txs {
+		if isBanTx(tx) {
+			var parsedBan model.BanTx
+			err := json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for _, tx := range processproposal.Txs[finishedBanTxIdx:] {
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+		}
+		if _, ok := bannedUsers[msg.Sender]; ok {
+			// sending us a tx from a banned user
+			return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+		}
+	}
+	return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_ACCEPT}, nil
+}
+
+// Deliver the decided block with its txs to the Application
+func (app *ForumApp) FinalizeBlock(_ context.Context, req *abci.RequestFinalizeBlock) (*abci.ResponseFinalizeBlock, error) {
+	fmt.Println("entered finalizeBlock")
+	// Iterate over Tx in current block
+	app.onGoingBlock = app.state.DB.GetDB().NewTransaction(true)
+	respTxs := make([]*types.ExecTxResult, len(req.Txs))
+	finishedBanTxIdx := len(req.Txs)
+	for i, tx := range req.Txs {
+		var err error
+
+		if isBanTx(tx) {
+			banTx := new(model.BanTx)
+			err = json.Unmarshal(tx, &banTx)
+			if err != nil {
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				err := UpdateOrSetUser(app.state.DB, banTx.UserName, true, app.onGoingBlock)
+				if err != nil {
+					panic(err)
+				}
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeOK}
+			}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for idx, tx := range req.Txs[finishedBanTxIdx:] {
+		msg, err := model.ParseMessage(tx)
+		i := idx + finishedBanTxIdx
+		if err != nil {
+			respTxs[i] = &types.ExecTxResult{Code: CodeTypeEncodingError}
+		} else {
+			// Check if this sender already existed; if not, add the user too
+			err := UpdateOrSetUser(app.state.DB, msg.Sender, false, app.onGoingBlock)
+			if err != nil {
+				panic(err)
+			}
+			// Add the message for this sender
+			message, err := model.AppendToExistingMsgs(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			app.onGoingBlock.Set([]byte(msg.Sender+"msg"), []byte(message))
+			chatHistory, err := model.AppendToChat(app.state.DB, *msg)
+			if err != nil {
+				panic(err)
+			}
+			// Append messages to chat history
+			app.onGoingBlock.Set([]byte("history"), []byte(chatHistory))
+			// This adds the user to the DB, but the data is not committed nor persisted until Comit is called
+			respTxs[i] = &types.ExecTxResult{Code: abci.CodeTypeOK}
+			app.state.Size++
+		}
+	}
+	app.state.Height = req.Height
```

Comment:
We should probably check that the request height is the app's height + 1, as a validation at the beginning of this function